File: tools/GhidraDecompile.java
// Ghidra post-script (Java) — descompila a C las funciones clave del Octatrack.
// Java siempre esta disponible en headless (Ghidra 12 no trae Jython por defecto).
// Fuerza la creacion de funcion en cada prologo hallado por string_func_map.py.
//@category Octatrack
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.address.Address;
import ghidra.program.model.address.AddressSpace;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.symbol.SourceType;
import ghidra.util.task.ConsoleTaskMonitor;

public class GhidraDecompile extends GhidraScript {
    @Override
    public void run() throws Exception {
        long[] targets = {
            0x40086d7aL, 0x4001fc1eL, 0x40069424L,
            0x400645ceL, 0x40022fdcL, 0x400867e0L
        };
        String[] labels = {
            "project_settings_serialize", "error_code_to_string", "format_card_handler",
            "save_project_handler", "collect_samples_handler", "project_sample_section"
        };

        AddressSpace sp = currentProgram.getAddressFactory().getDefaultAddressSpace();
        FunctionManager fm = currentProgram.getFunctionManager();
        DecompInterface dec = new DecompInterface();
        dec.openProgram(currentProgram);
        ConsoleTaskMonitor mon = new ConsoleTaskMonitor();

        for (int i = 0; i < targets.length; i++) {
            Address a = sp.getAddress(targets[i]);
            Function f = fm.getFunctionAt(a);
            if (f == null) {
                try { disassemble(a); f = createFunction(a, labels[i]); }
                catch (Exception e) { println("[!] no pude crear funcion @ " + a + ": " + e); }
            }
            if (f == null) { println("==== " + labels[i] + ": sin funcion @ " + a + " ===="); continue; }
            try { f.setName(labels[i], SourceType.USER_DEFINED); } catch (Exception e) {}

            DecompileResults res = dec.decompileFunction(f, 120, mon);
            println("\n==================== " + labels[i] + " @ " + a + " ====================");
            if (res != null && res.decompileCompleted()) {
                println(res.getDecompiledFunction().getC());
            } else {
                println("  (descompilacion fallida: " + (res != null ? res.getErrorMessage() : "null") + ")");
            }
        }
        println("\n[GhidraDecompile] fin.");
    }
}

File: tools/ghidra/ghidra_decompile.py
# Ghidra post-script — descompila a C las funciones clave del Octatrack.
# Se ejecuta via analyzeHeadless tras la auto-anotacion. Fuerza la creacion de
# funcion en cada direccion objetivo (prologos hallados por string_func_map.py) y
# vuelca el pseudo-C.
#
# @category Octatrack
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
from ghidra.program.model.symbol import SourceType

# (vaddr, etiqueta) — funciones ancladas a strings de UI (ver out/string_function_map.txt)
TARGETS = [
    (0x40086d7a, "project_settings_serialize"),  # 58 claves de config del proyecto
    (0x4001fc1e, "error_code_to_string"),         # familia 'ERROR CODE: %d'
    (0x40069424, "format_card_handler"),          # 'FORMAT CARD'
    (0x400645ce, "save_project_handler"),         # 'SAVE PROJECT'
    (0x40022fdc, "collect_samples_handler"),      # 'COLLECT SAMPLES'
    (0x400867e0, "project_sample_section"),       # token '[SAMPLE]'
]

af = currentProgram.getAddressFactory().getDefaultAddressSpace()  # noqa: F821
fm = currentProgram.getFunctionManager()                          # noqa: F821

dec = DecompInterface()
dec.openProgram(currentProgram)  # noqa: F821
mon = ConsoleTaskMonitor()

for vaddr, label in TARGETS:
    a = af.getAddress(vaddr)
    f = fm.getFunctionAt(a)
    if f is None:
        try:
            disassemble(a)                       # noqa: F821
            f = createFunction(a, label)         # noqa: F821
        except Exception as e:
            print("[!] no pude crear funcion en 0x%x: %s" % (vaddr, e))
    if f is None:
        print("==== 0x%08x (%s): sin funcion ====" % (vaddr, label))
        continue
    try:
        f.setName(label, SourceType.USER_DEFINED)
    except Exception:
        pass
    res = dec.decompileFunction(f, 90, mon)
    print("\n==================== %s @ 0x%08x ====================" % (label, vaddr))
    if res and res.decompileCompleted():
        print(res.getDecompiledFunction().getC())
    else:
        print("  (descompilacion fallida: %s)" %
              (res.getErrorMessage() if res else "sin resultado"))

print("\n[ghidra_decompile] fin.")
